pkg/descheduler/strategies/topologyspreadconstraint.go

```diff
@@ -0,0 +1,336 @@
+			i++
+			continue
+		}
+
+		// remove pods from the higher topology and add them to the list of pods to be evicted
+		// also (just for tracking), add them to the list of pods in the lower topology
+		aboveToEvict := sortedDomains[j].pods[len(sortedDomains[j].pods)-movePods:]
+		for i := range aboveToEvict {
+			// if the pod has a hard nodeAffinity or nodeSelector, don't bother evicting it as it will just end up back on the same node
+			// however we still account for it "being evicted" so the algorithm can complete
+			if aboveToEvict[i].Spec.NodeSelector != nil ||
+				(aboveToEvict[i].Spec.Affinity != nil &&
+					aboveToEvict[i].Spec.Affinity.NodeAffinity != nil &&
+					aboveToEvict[i].Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil) {
+				continue
+			}
+			podsForEviction[aboveToEvict[i]] = struct{}{}
+		}
+		sortedDomains[j].pods = sortedDomains[j].pods[:len(sortedDomains[j].pods)-movePods]
+		sortedDomains[i].pods = append(sortedDomains[i].pods, aboveToEvict...)
+	}
+}
+
+// sortDomains sorts and splits the list of topology domains based on their size
+// it also sorts the list of pods within the domains based on their node affinity/selector and priority
+func sortDomains(constraintTopologyPairs map[topologyPair][]*v1.Pod, isEvictable func(*v1.Pod) bool) []topology {
+	sortedTopologies := make([]topology, 0, len(constraintTopologyPairs))
+	// sort the topologies and return 2 lists: those <= the average and those > the average (> list inverted)
+	for pair, list := range constraintTopologyPairs {
+		// Sort the pods within the domain so that the lowest priority pods are considered first for eviction,
+		// followed by the highest priority,
+		// followed by the lowest priority pods with affinity or nodeSelector,
+		// followed by the highest priority pods with affinity or nodeSelector
+		sort.Slice(list, func(i, j int) bool {
+			// any non-evictable pods should be considered last (ie, first in the list)
+			if !isEvictable(list[i]) || !isEvictable(list[j]) {
+				// false - i is the only non-evictable, so return true to put it first
+				// true - j is non-evictable, so return false to put j before i
+				// if true and both and non-evictable, order doesn't matter
+				return !isEvictable(list[j])
+			}
+
+			// if both pods have selectors/affinity, compare them by their priority
+			if hasSelectorOrAffinity(*list[i]) && hasSelectorOrAffinity(*list[j]) {
+				return comparePodsByPriority(list[i], list[j])
+			} else if hasSelectorOrAffinity(*list[i]) && !hasSelectorOrAffinity(*list[j]) {
+				// if just one has selectors/affinity
+				// i should come first
+				return true
+			} else if !hasSelectorOrAffinity(*list[i]) && hasSelectorOrAffinity(*list[j]) {
+				// if just one has selectors/affinity
+				// j should come first
+				return false
+			} else {
+				// if neither has selectors/affinity, fall back to just priority
+				return comparePodsByPriority(list[i], list[j])
+			}
+		})
```

Comment:
@lixiang233 or @ingvagabund mind double checking this sorting code to make sure it's right? The goal is:

1. Put non-evictable pods first
2. Put pods with affinity or selector next
3. Order remaining pods by priority